File: leetcode/easy/MajorityElement.py
from typing import List


class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        """
        Finds the majority element in an array.
        The majority element is the element that appears more than ⌊n / 2⌋ times.
        This problem assumes that the majority element always exists in the array.

        This solution uses the Boyer-Moore Voting Algorithm, which is a highly
        efficient O(n) time and O(1) space algorithm for this problem.

        Algorithm Intuition:
        The algorithm works by maintaining a `candidate` for the majority element and a `count`.
        As we iterate through the array:
        - If the count is zero, we pick the current element as the new candidate.
        - If the current element is the same as the candidate, we increment the count.
        - If it's different, we decrement the count.

        The key idea is that if an element truly is the majority, its "votes" (increments)
        will outnumber the "votes" against it (decrements from other elements), ensuring
        it remains the candidate at the end.
        """
        candidate = None
        count = 0

        for num in nums:
            # If count is 0, we choose the current element as the new candidate.
            if count == 0:
                candidate = num

            # Increment count if the element matches the candidate, otherwise decrement.
            count += 1 if num == candidate else -1

        # Since the problem guarantees a majority element exists, the final
        # candidate will be the majority element.
        return candidate

File: leetcode/easy/MajorityElement.java
/**
 * Solution for LeetCode problem "Majority Element".
 * This class provides a method to find the majority element in an array.
 * The majority element is the element that appears more than ⌊n / 2⌋ times.
 * It is assumed that the array is non-empty and the majority element always exists in the array.
 */
class Solution {

    /**
     * Finds the majority element in the given array using Boyer-Moore Voting Algorithm.
     *
     * Boyer-Moore Voting Algorithm works on the principle that if an element is the majority element,
     * its count will remain positive even after canceling out with all other non-majority elements.
     *
     * The algorithm maintains a `candidate` for the majority element and a `count`.
     * - Initialize `count = 0` and `candidate = null`.
     * - Iterate through the array `nums`:
     *   - If `count` is 0, set the current element as the new `candidate`.
     *   - If the current element is the same as `candidate`, increment `count`.
     *   - If the current element is different from `candidate`, decrement `count`.
     * - After iterating through the entire array, the `candidate` will be the majority element.
     *
     * @param nums The array of integers.
     * @return The majority element in the array.
     */
    public int majorityElement(int[] nums) {
        // Initialize candidate and count for Boyer-Moore Voting Algorithm.
        Integer candidate = null;
        int count = 0;

        // Iterate through each number in the array.
        for (int num : nums) {
            // If count is 0, set the current number as the new candidate.
            // This means the previous candidate's votes have been cancelled out.
            if (count == 0) {
                candidate = num;
            }

            // If the current number is the same as the candidate, increment count.
            // Otherwise, decrement count.
            count += (num == candidate) ? 1 : -1;
        }

        // The candidate at the end of the iteration will be the majority element.
        // The problem statement guarantees that a majority element always exists.
        return candidate;
    }
}
